docs/blocks/states.md

```diff
@@ -1,96 +1,149 @@
 
 ```java
-this.registerDefaultState(
-  this.stateDefinition.any()
-    .setValue(FACING, Direction.NORTH)
-    .setValue(OPEN, false)
-    .setValue(HINGE, DoorHingeSide.LEFT)
-    .setValue(POWERED, false)
-    .setValue(HALF, DoubleBlockHalf.LOWER)
-);
+leverBlockState = leverBlockState.setValue(LeverBlock.FACING, Direction.SOUTH);
 ```
 
-If you wish to change what `BlockState` is used when placing your block, you can overwrite `Block#getStateForPlacement(BlockPlaceContext)`. This can be used to, for example, set the direction of your block depending on where the player is standing when they place it.
+:::note
+`BlockState`s are immutable. This means that when you call `#setValue(Property<T>, T)`, you are not actually modifying the blockstate. Instead, a lookup is performed internally, and you are given the blockstate object you requested, which is the one and only object that exists with these exact property values. This also means that just calling `state#setValue` without saving it into a variable (for example back into `state`) does nothing.
+:::
+
+To get a `BlockState` from the level, use `Level#getBlockState(BlockPos)`.
+
+### `Level#setBlock`
 
-Because `BlockState`s are immutable, and all combinations of their properties are generated on startup of the game, calling `BlockState#setValue(Property<T>, T)` will simply go to the `Block`'s `StateHolder` and request the `BlockState` with the set of values you want.
+To set a `BlockState` in the level, use `Level#setBlock(BlockPos, BlockState, int)`.
 
-Because all possible `BlockState`s are generated at startup, you are free and encouraged to use the reference equality operator (`==`) to check if two `BlockState`s are equal.
+The `int` parameter deserves some extra explanation, as its meaning is not immediately obvious. It denotes what is known as update flags.
 
-Using `BlockState`'s
----------------------
+To help setting the update flags correctly, there are a number of `int` constants in `Block`, prefixed with `UPDATE_`. These constants can be bitwise-ORed together (for example `Block.UPDATE_NEIGHBORS | Block.UPDATE_CLIENTS`) if you wish to combine them.
 
-You can get the value of a property by calling `BlockState#getValue(Property<?>)`, passing it the property you want to get the value of.
-If you want to get a `BlockState` with a different set of values, simply call `BlockState#setValue(Property<T>, T)` with the property and its value.
+- `Block.UPDATE_NEIGHBORS` sends an update to the neighboring blocks. More specifically, it calls `Block#neighborChanged`, which calls a number of methods, most of which are redstone-related in some way.
+- `Block.UPDATE_CLIENTS` syncs the block update to the client.
+- `Block.UPDATE_INVISIBLE` explicitly does not update. This also overrules `Block.UPDATE_CLIENTS`, causing the update to not be synced.
+- `Block.UPDATE_IMMEDIATE` forces the update immediately, instead of on the next tick.
```

Comment:
Sort of? This one is a bit hard to explain. Just because this is updated immediately doesn't mean that the changes will be visibly applied on the same tick. You might want to take a gander through the method in question again.

Reply:
This is well-hidden in LevelRenderer#blockChanged, which is why I couldn't find it.